src/modules/flow/http-server/http-server.c

```diff
@@ -214,125 +203,131 @@ boolean_close(struct sol_flow_node *node, void *data)
+common_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
 {
     int r;
     struct http_data *mdata = data;
     struct sol_flow_node_type_http_server_boolean_options *opts =
         (struct sol_flow_node_type_http_server_boolean_options *)options;
 
-    r = start_server(mdata, opts->path, boolean_response_cb);
+    r = start_server(mdata, node, opts->path);
     SOL_INT_CHECK(r, < 0, r);
 
     mdata->value.b = opts->value;
-    mdata->node = node;
 
     return 0;
 }
 
+static int
+common_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id,
+    const struct sol_flow_packet *packet)
+{
+    struct http_data *mdata = data;
+    const struct http_server_node_type *type;
+    int r;
+
+    type = (const struct http_server_node_type *)
+        sol_flow_node_get_type(node);
+
+    r = type->process_cb(mdata, packet);
+    SOL_INT_CHECK(r, < 0, r);
+
+    r = sol_http_server_set_last_modified(server, mdata->path, time(NULL));
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+static int
+boolean_post_cb(struct http_data *mdata, struct sol_flow_node *node,
+    struct sol_http_param_value *value)
+{
+    if (streq(value->value.key_value.value, "true"))
+        mdata->value.b = true;
+    else if (streq(value->value.key_value.value, "false"))
+        mdata->value.b = false;
+    else
+        return -EINVAL;
+
+    sol_flow_send_boolean_packet(node,
+        SOL_FLOW_NODE_TYPE_HTTP_SERVER_BOOLEAN__OUT__OUT,
+        mdata->value.b);
+
+    return 0;
+}
+
+static int
+boolean_response_cb(struct http_data *mdata, struct sol_buffer *content, bool json)
+{
+    int r;
+
+    r = sol_buffer_append_printf(content, "%s", mdata->value.b == true ? "true" : "false");
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+static int
+boolean_process_cb(struct http_data *mdata, const struct sol_flow_packet *packet)
+{
+    return sol_flow_packet_get_boolean(packet, &mdata->value.b);
+}
+
 /* ------------------------------------- string ------------------------------------------- */
 
 static int
-string_response_cb(void *data, struct sol_http_request *request)
+string_response_cb(struct http_data *mdata, struct sol_buffer *content, bool json)
 {
     int r = 0;
-    uint16_t idx;
-    bool send_json = false;
-    enum sol_http_method method;
-    struct http_data *mdata = data;
-    char *response_str;
-    struct sol_http_param_value *value;
-    struct sol_http_response response = {
-        .api_version = SOL_HTTP_RESPONSE_API_VERSION,
-        .content = SOL_BUFFER_INIT_EMPTY,
-        .param = SOL_HTTP_REQUEST_PARAM_INIT,
-        .response_code = SOL_HTTP_STATUS_OK
-    };
 
-    method = sol_http_request_get_method(request);
-    response.url = sol_http_request_get_url(request);
+    if (json) {
+        size_t escaped_len = sol_json_calculate_escaped_string_len(mdata->value.s);
+        r = sol_buffer_ensure(content, content->used + escaped_len);
```

Comment:
it will, but you ensure `escaped_len`,  then you consume one char with `append_char()`, then you have `escaped_len - 1` available, then you:

```c
sol_json_escape_string(mdata->value.s, sol_buffer_at_end(content), escaped_len);
```
yet you don't have `escaped_len` anymore.